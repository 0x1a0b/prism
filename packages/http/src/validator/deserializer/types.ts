import { ISchema } from '@stoplight/types/schema';

export type DeserializeHttpHeader = (value: string, type: string, explode: boolean) => any;

export interface IHttpHeaderParamStyleDeserializer {
  supports: (style: string) => boolean;
  deserialize: DeserializeHttpHeader;
}

export interface IHttpParamDeserializerRegistry<T extends Function> {
  get(style: string): T | undefined;
}

export interface IHttpParamStyleDeserializer<T> {
  supports: (style: string) => boolean;
  deserialize: T;
}

export type DeserializeHttpQuery = (
  key: string,
  query: {
    [name: string]: string | string[];
  },
  schema: ISchema,
  explode: boolean
) => any;

export interface IHttpQueryParamStyleDeserializer
  extends IHttpParamStyleDeserializer<DeserializeHttpQuery> {
  deserialize: DeserializeHttpQuery;
}
