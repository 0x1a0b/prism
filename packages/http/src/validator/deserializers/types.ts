import { HttpParamStyles, ISchema } from '@stoplight/types';

import { IHttpNameValue, IHttpNameValues } from '../../types';

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: (name: string, parameters: Parameters, schema: ISchema, explode?: boolean) => any;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
