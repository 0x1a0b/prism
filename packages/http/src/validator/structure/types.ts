import { IValidation } from '@stoplight/prism-core';

export interface IHttpValidator<Obj, Specs> {
  validate(obj: Obj, specs: Specs[], mediaType?: string): IValidation[];
}
