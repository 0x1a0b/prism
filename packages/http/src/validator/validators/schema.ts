import { IValidation, ValidationSeverity } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types';
import * as Ajv from 'ajv';
import { Ajv as AjvClass } from 'ajv';

import { ISchemaValidator } from './types';
import { convertAjvErrors } from './utils';

const SUPPORTED_MEDIATYPES = ['application/json'];

export class JSONSchemaValidator implements ISchemaValidator<ISchema> {
  private ajv: AjvClass;

  constructor() {
    this.ajv = new Ajv({ allErrors: true, messages: true });
  }

  public validate(content: any, schema: ISchema): IValidation[] {
    const validate = this.ajv.compile(schema);
    const errors = validate(JSON.parse(content)) ? [] : validate.errors;
    return convertAjvErrors(errors, ValidationSeverity.ERROR);
  }

  public supports(mediaType: string) {
    return SUPPORTED_MEDIATYPES.indexOf(mediaType) > -1;
  }
}
