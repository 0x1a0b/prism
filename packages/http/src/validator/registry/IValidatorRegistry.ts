import { IValidation } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types/schema';

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: ISchema) => IValidation[]) | undefined;
}
