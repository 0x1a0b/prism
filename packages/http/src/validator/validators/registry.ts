import { IValidation } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types/schemas';

import { ISchemaValidator, IValidatorRegistry } from './types';

export class ValidatorRegistry implements IValidatorRegistry {
  constructor(private validators: Array<ISchemaValidator<ISchema>>) {}

  public get(mediaType: string): ((content: any, schema: ISchema) => IValidation[]) | undefined {
    const validator = this.validators.find(v => v.supports(mediaType));

    if (!validator) {
      return;
    }

    return validator.validate.bind(validator);
  }
}
