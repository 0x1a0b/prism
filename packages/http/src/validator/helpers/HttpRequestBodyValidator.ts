import { IValidation } from '@stoplight/prism-core';
import { IValidatorRegistry } from '@stoplight/prism-http/validator/registry/IValidatorRegistry';
import { IHttpContent, IHttpRequest, IHttpRequestBody } from '@stoplight/types/http';

export class HttpRequestBodyValidator {
  constructor(private validatorRegistry: IValidatorRegistry) {}

  public validate(body: any, requestSpec?: IHttpRequest, mediaType?: string): IValidation[] {
    if (!requestSpec) {
      return [];
    }

    const content = this.getContent(requestSpec.body, mediaType);

    if (!content) {
      return [];
    }

    if (!content.schema) {
      return [];
    }

    const validate = this.validatorRegistry.get(content.mediaType);

    if (!validate) {
      return [];
    }

    return validate(body, content.schema);
  }

  private getContent(
    requestBodySpec?: IHttpRequestBody,
    mediaType?: string
  ): IHttpContent | undefined {
    if (!requestBodySpec) {
      return;
    }

    const contentList = requestBodySpec.content;

    if (!mediaType) {
      return contentList[0];
    }

    const content = contentList.find(c => c.mediaType === mediaType);

    if (!content) {
      return contentList[0];
    }

    return content;
  }

}
