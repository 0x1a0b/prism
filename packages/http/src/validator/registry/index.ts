import { JSONSchemaValidator } from '@stoplight/prism-http/validator/registry/JSONSchemaValidator';
import { ValidatorRegistry } from '@stoplight/prism-http/validator/registry/ValidatorRegistry';

export const validatorRegistry = new ValidatorRegistry([new JSONSchemaValidator()]);
