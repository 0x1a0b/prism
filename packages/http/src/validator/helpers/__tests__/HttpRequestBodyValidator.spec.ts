import { ValidationSeverity } from '@stoplight/prism-core/types';
import { IValidatorRegistry } from '@stoplight/prism-http/validator/registry/IValidatorRegistry';
import { ISchema } from '@stoplight/types/schema';
import { HttpRequestBodyValidator } from '../HttpRequestBodyValidator';

describe('HttpRequestBodyValidator', () => {
  const validatorRegistry = { get: () => () => [] } as IValidatorRegistry;
  const httpRequestBodyValidator = new HttpRequestBodyValidator(validatorRegistry);

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(validatorRegistry, 'get');
  });

  describe('validate()', () => {
    describe('request spec is not set', () => {
      it('returns no validation errors', () => {
        expect(httpRequestBodyValidator.validate('test', undefined, 'application/json')).toEqual(
          []
        );
      });
    });

    describe('body spec is not set', () => {
      it('returns no validation errors', () => {
        expect(httpRequestBodyValidator.validate('test', {}, 'application/json')).toEqual([]);
      });
    });

    describe('body spec is set', () => {
      describe('request media type is not provided', () => {
        it('returns no validation errors', () => {
          expect(
            httpRequestBodyValidator.validate('test', {
              body: {
                content: [{ mediaType: 'application/not-exists-son' }],
              },
            })
          ).toEqual([]);
        });
      });

      describe('request media type was not found in spec', () => {
        it('returns no validation errors', () => {
          expect(
            httpRequestBodyValidator.validate(
              'test',
              {
                body: {
                  content: [{ mediaType: 'application/not-exists-son' }],
                },
              },
              'application/json'
            )
          ).toEqual([]);
        });
      });

      describe('body schema is provided', () => {
        describe('validator for given media type does not exists', () => {
          it('return no validation errors', () => {
            jest.spyOn(validatorRegistry, 'get').mockImplementation(mediaType => {
              expect(mediaType).toBe('application/json');
              return;
            });

            expect(
              httpRequestBodyValidator.validate(
                'test',
                {
                  body: {
                    content: [{ mediaType: 'application/json', schema: {} }],
                  },
                },
                'application/json'
              )
            ).toEqual([]);

            expect(validatorRegistry.get).toHaveBeenCalledTimes(1);
          });
        });

        describe('validator for given media type exists', () => {
          it('return no validation errors', () => {
            const mockSchema = { type: 'string' };

            jest.spyOn(validatorRegistry, 'get').mockImplementation(mediaType => {
              expect(mediaType).toBe('application/json');
              return (content: any, schema: ISchema) => {
                expect(content).toEqual('test');
                expect(schema).toEqual(mockSchema);
                return [
                  {
                    path: ['a'],
                    name: 'testName',
                    summary: 'testSummary',
                    severity: ValidationSeverity.ERROR,
                    message: 'testMessage',
                  },
                ];
              };
            });

            expect(
              httpRequestBodyValidator.validate(
                'test',
                {
                  body: {
                    content: [{ mediaType: 'application/json', schema: mockSchema }],
                  },
                },
                'application/json'
              )
            ).toMatchSnapshot();

            expect(validatorRegistry.get).toHaveBeenCalledTimes(1);
          });
        });
      });
    });

    it('', () => {
      httpRequestBodyValidator.validate(
        'test',
        {
          body: {
            content: [
              {
                mediaType: 'application/json',
              },
            ],
          },
        },
        'application/json'
      );
    });
  });
});
