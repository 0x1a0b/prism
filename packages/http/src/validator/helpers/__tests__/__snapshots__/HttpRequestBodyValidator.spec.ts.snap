// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpRequestBodyValidator validate() body spec is set body schema is provided validator for given media type exists return no validation errors 1`] = `
Array [
  Object {
    "message": "testMessage",
    "name": "testName",
    "path": Array [
      "a",
    ],
    "severity": "error",
    "summary": "testSummary",
  },
]
`;
